lib/Backend/GlobOpt.cpp

```diff
@@ -19198,30 +19198,54 @@ GlobOpt::OptDstIsInvariant(IR::RegOpnd *dst)
 }
 
 void
-GlobOpt::OptHoistToLandingPadUpdateValueType(
-    BasicBlock* landingPad,
+GlobOpt::OptHoistUpdateValueType(
+    Loop* loop,
     IR::Instr* instr,
-    IR::Opnd* opnd,
+    IR::Opnd* srcOpnd,
     Value* opndVal)
 {
-    if (instr->m_opcode == Js::OpCode::FromVar)
+    if (opndVal == nullptr || instr->m_opcode == Js::OpCode::FromVar)
     {
         return;
     }
 
-    Sym* opndSym = opnd->GetSym();;
+    Sym* opndSym = srcOpnd->GetSym();;
 
     if (opndSym)
     {
-        if (opndVal == nullptr)
-        {
-            opndVal = FindValue(opndSym);
-        }
-
+        BasicBlock* landingPad = loop->landingPad;
         Value* opndValueInLandingPad = FindValue(landingPad->globOptData.symToValueMap, opndSym);
         Assert(opndVal->GetValueNumber() == opndValueInLandingPad->GetValueNumber());
 
-        opnd->SetValueType(opndValueInLandingPad->GetValueInfo()->Type());
+        ValueType opndValueTypeInLandingPad = opndValueInLandingPad->GetValueInfo()->Type();
+
+        if (srcOpnd->GetValueType() != opndValueTypeInLandingPad)
+        {
+            if (instr->m_opcode == Js::OpCode::SetConcatStrMultiItemBE)
```

Comment:
We are not checking if the Conv_PrimStr instr is actually required, are we? I mean, if the bailout ensuring the string type was also hoisted, we wouldn't need conv_primstr.